### PR TITLE
executor: fix data race in the TestHistoryReadInTxn

### LINKED ISSUE #34666: DATA RACE in the TestHistoryReadInTxn
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
==================
WARNING: DATA RACE
Write at 0x00c0089bb6d0 by goroutine 66:
  github.com/pingcap/tidb/sessionctx/variable.(*SessionVars).InitStatementContext()
      /go/tidb/sessionctx/variable/session.go:1045 +0x20d
  github.com/pingcap/tidb/executor.ResetContextOfStmt()
      /go/tidb/executor/executor.go:1769 +0x15a
  github.com/pingcap/tidb/session.(*session).ExecuteStmt()
      /go/tidb/session/session.go:1881 +0x37c
  github.com/pingcap/tidb/testkit.(*TestKit).Exec()
      /go/tidb/testkit/testkit.go:214 +0x701
  github.com/pingcap/tidb/testkit.(*TestKit).MustExec()
      /go/tidb/testkit/testkit.go:88 +0x99
  github.com/pingcap/tidb/executor_test.TestHistoryReadInTxn.func2()
      /go/tidb/executor/executor_test.go:5463 +0xca6
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1439 +0x213
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1486 +0x47
Previous write at 0x00c0089bb6d0 by goroutine 154:
  sync/atomic.AddInt32()
      /usr/local/go/src/runtime/race_amd64.s:279 +0xb
  sync/atomic.AddInt32()
      <autogenerated>:1 +0x1a
  github.com/pingcap/tidb/util/memory.(*Tracker).getParent.func1()
      /go/tidb/util/memory/tracker.go:556 +0x39
  runtime.deferreturn()
      /usr/local/go/src/runtime/panic.go:436 +0x32
  github.com/pingcap/tidb/util/memory.(*Tracker).Consume()
      /go/tidb/util/memory/tracker.go:346 +0x84
  github.com/pingcap/tidb/store/copr.(*copIteratorWorker).sendToRespCh()
      /go/tidb/store/copr/coprocessor.go:553 +0x1a9
  github.com/pingcap/tidb/store/copr.(*copIteratorWorker).handleCopResponse()
      /go/tidb/store/copr/coprocessor.go:1025 +0x237c
  github.com/pingcap/tidb/store/copr.(*copIteratorWorker).handleTaskOnce()
      /go/tidb/store/copr/coprocessor.go:792 +0x1a74
  github.com/pingcap/tidb/store/copr.(*copIteratorWorker).handleTask()
      /go/tidb/store/copr/coprocessor.go:676 +0x1e5
  github.com/pingcap/tidb/store/copr.(*copIteratorWorker).run()
      /go/tidb/store/copr/coprocessor.go:418 +0x173
  github.com/pingcap/tidb/store/copr.(*copIterator).open.func1()
      /go/tidb/store/copr/coprocessor.go:450 +0x58
Goroutine 66 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1486 +0x724
  github.com/pingcap/tidb/executor_test.TestHistoryReadInTxn()
      /go/tidb/executor/executor_test.go:5422 +0x598
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1439 +0x213
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1486 +0x47
Goroutine 154 (finished) created at:
  github.com/pingcap/tidb/store/copr.(*copIterator).open()
      /go/tidb/store/copr/coprocessor.go:450 +0xc4
  github.com/pingcap/tidb/store/copr.(*CopClient).Send()
      /go/tidb/store/copr/coprocessor.go:143 +0x10c9
  github.com/pingcap/tidb/distsql.Select()
      /go/tidb/distsql/distsql.go:100 +0x8c3
  github.com/pingcap/tidb/distsql.SelectWithRuntimeStats()
      /go/tidb/distsql/distsql.go:149 +0xbb
  github.com/pingcap/tidb/executor.selectResultHook.SelectResult()
      /go/tidb/executor/table_reader.go:53 +0x204
  github.com/pingcap/tidb/executor.(*TableReaderExecutor).buildResp()
      /go/tidb/executor/table_reader.go:310 +0x855
  github.com/pingcap/tidb/executor.(*TableReaderExecutor).Open()
      /go/tidb/executor/table_reader.go:210 +0x158b
  github.com/pingcap/tidb/executor.(*baseExecutor).Open()
      /go/tidb/executor/executor.go:199 +0x101
  github.com/pingcap/tidb/executor.(*SelectLockExec).Open()
      /go/tidb/executor/executor.go:1001 +0x199
  github.com/pingcap/tidb/executor.(*ExecStmt).Exec()
      /go/tidb/executor/adapter.go:420 +0x9de
  github.com/pingcap/tidb/session.runStmt()
      /go/tidb/session/session.go:2066 +0x6cb
  github.com/pingcap/tidb/session.(*session).ExecuteStmt()
      /go/tidb/session/session.go:1935 +0xdba
  github.com/pingcap/tidb/testkit.(*TestKit).Exec()
      /go/tidb/testkit/testkit.go:214 +0x701
  github.com/pingcap/tidb/testkit.(*TestKit).ExecToErr()
      /go/tidb/testkit/testkit.go:251 +0x6e
  github.com/pingcap/tidb/executor_test.TestHistoryReadInTxn.func2()
      /go/tidb/executor/executor_test.go:5446 +0x1204
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1439 +0x213
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1486 +0x47
==================
==================
WARNING: DATA RACE
Write at 0x00c0089bb6d8 by goroutine 66:
  github.com/pingcap/tidb/sessionctx/variable.(*SessionVars).InitStatementContext()
      /go/tidb/sessionctx/variable/session.go:1045 +0x20d
  github.com/pingcap/tidb/executor.ResetContextOfStmt()
      /go/tidb/executor/executor.go:1769 +0x15a
  github.com/pingcap/tidb/session.(*session).ExecuteStmt()
      /go/tidb/session/session.go:1881 +0x37c
  github.com/pingcap/tidb/testkit.(*TestKit).Exec()
      /go/tidb/testkit/testkit.go:214 +0x701
  github.com/pingcap/tidb/testkit.(*TestKit).MustExec()
      /go/tidb/testkit/testkit.go:88 +0x99
  github.com/pingcap/tidb/executor_test.TestHistoryReadInTxn.func2()
      /go/tidb/executor/executor_test.go:5463 +0xca6
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1439 +0x213
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1486 +0x47
Previous read at 0x00c0089bb6d8 by goroutine 154:
  github.com/pingcap/tidb/util/memory.(*Tracker).getParent()
      /go/tidb/util/memory/tracker.go:557 +0x94
  github.com/pingcap/tidb/util/memory.(*Tracker).Consume()
      /go/tidb/util/memory/tracker.go:346 +0x84
  github.com/pingcap/tidb/store/copr.(*copIteratorWorker).sendToRespCh()
      /go/tidb/store/copr/coprocessor.go:553 +0x1a9
  github.com/pingcap/tidb/store/copr.(*copIteratorWorker).handleCopResponse()
      /go/tidb/store/copr/coprocessor.go:1025 +0x237c
  github.com/pingcap/tidb/store/copr.(*copIteratorWorker).handleTaskOnce()
      /go/tidb/store/copr/coprocessor.go:792 +0x1a74
  github.com/pingcap/tidb/store/copr.(*copIteratorWorker).handleTask()
      /go/tidb/store/copr/coprocessor.go:676 +0x1e5
  github.com/pingcap/tidb/store/copr.(*copIteratorWorker).run()
      /go/tidb/store/copr/coprocessor.go:418 +0x173
  github.com/pingcap/tidb/store/copr.(*copIterator).open.func1()
      /go/tidb/store/copr/coprocessor.go:450 +0x58
Goroutine 66 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1486 +0x724
  github.com/pingcap/tidb/executor_test.TestHistoryReadInTxn()
      /go/tidb/executor/executor_test.go:5422 +0x598
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1439 +0x213
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1486 +0x47
Goroutine 154 (finished) created at:
  github.com/pingcap/tidb/store/copr.(*copIterator).open()
      /go/tidb/store/copr/coprocessor.go:450 +0xc4
  github.com/pingcap/tidb/store/copr.(*CopClient).Send()
      /go/tidb/store/copr/coprocessor.go:143 +0x10c9
  github.com/pingcap/tidb/distsql.Select()
      /go/tidb/distsql/distsql.go:100 +0x8c3
  github.com/pingcap/tidb/distsql.SelectWithRuntimeStats()
      /go/tidb/distsql/distsql.go:149 +0xbb
  github.com/pingcap/tidb/executor.selectResultHook.SelectResult()
      /go/tidb/executor/table_reader.go:53 +0x204
  github.com/pingcap/tidb/executor.(*TableReaderExecutor).buildResp()
      /go/tidb/executor/table_reader.go:310 +0x855
  github.com/pingcap/tidb/executor.(*TableReaderExecutor).Open()
      /go/tidb/executor/table_reader.go:210 +0x158b
  github.com/pingcap/tidb/executor.(*baseExecutor).Open()
      /go/tidb/executor/executor.go:199 +0x101
  github.com/pingcap/tidb/executor.(*SelectLockExec).Open()
      /go/tidb/executor/executor.go:1001 +0x199
  github.com/pingcap/tidb/executor.(*ExecStmt).Exec()
      /go/tidb/executor/adapter.go:420 +0x9de
  github.com/pingcap/tidb/session.runStmt()
      /go/tidb/session/session.go:2066 +0x6cb
  github.com/pingcap/tidb/session.(*session).ExecuteStmt()
      /go/tidb/session/session.go:1935 +0xdba
  github.com/pingcap/tidb/testkit.(*TestKit).Exec()
      /go/tidb/testkit/testkit.go:214 +0x701
  github.com/pingcap/tidb/testkit.(*TestKit).ExecToErr()
      /go/tidb/testkit/testkit.go:251 +0x6e
  github.com/pingcap/tidb/executor_test.TestHistoryReadInTxn.func2()
      /go/tidb/executor/executor_test.go:5446 +0x1204
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1439 +0x213
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1486 +0x47
================== 
```

### PR DESCRIPTION
**Pull request by hawkingrei**
Signed-off-by: Weizhen Wang <wangweizhen@pingcap.com>

### What problem does this PR solve?

Issue Number: close #34666

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [x] Integration test

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by lcwangchao — APPROVED** [review]

File: executor/executor_test.go [hunk 1]
```diff
@@ -5441,41 +5443,42 @@ func TestHistoryReadInTxn(t *testing.T) {
 					}
 
 					for _, sql := range notAllowedSQLs {
-						err := tk.ExecToErr(sql)
+						err := tkt.ExecToErr(sql)
 						require.Errorf(t, err, "can not execute write statement when 'tidb_snapshot' is set")
+						time.Sleep(20 * time.Millisecond)
```
Comment (tiancaiamao) [inline executor/executor_test.go:5448]: The real problem of this data race, is that the query result set is not closed.
The leaked coprocessor goroutine still working in the background, and then race with the next SQL statement on `session.SessionVars`

It's fixed accidently by this PR here https://github.com/pingcap/tidb/pull/34837/files#diff-cda21674d674295dde958bfdfe5d6ecd21ae07fd310fa907920e1214108d974fR596


Sleep 20ms is not the correct workaround, it just reduce the change of DATA RACE

**Pull request closed without merging**